back/src/Model/Space.ts

```diff
@@ -0,0 +1,139 @@
+            }
+            if (spaceUser.hasColor()) {
+                user.setColor(spaceUser.getColor()?.getValue() as string);
+            }
+            if (spaceUser.getCharacterlayersList().length > 0) {
+                user.setCharacterlayersList(spaceUser.getCharacterlayersList());
+            }
+            if (spaceUser.hasIslogged()) {
+                user.setIslogged(spaceUser.getIslogged()?.getValue() as boolean);
+            }
+            if (spaceUser.hasAvailabilitystatus()) {
+                user.setAvailabilitystatus(spaceUser.getAvailabilitystatus()?.getValue() as number);
+            }
+            if (spaceUser.hasRoomname()) {
+                user.setRoomname(spaceUser.getRoomname());
+            }
+            if (spaceUser.hasVisitcardurl()) {
+                user.setVisitcardurl(spaceUser.getVisitcardurl());
+            }
+            if (spaceUser.hasScreensharing()) {
+                user.setScreensharing(spaceUser.getScreensharing()?.getValue() as boolean);
+            }
+            if (spaceUser.hasAudiosharing()) {
+                user.setAudiosharing(spaceUser.getAudiosharing()?.getValue() as boolean);
+            }
+            if (spaceUser.hasVideosharing()) {
+                user.setVideosharing(spaceUser.getVideosharing()?.getValue() as boolean);
+            }
+            usersList.set(spaceUser.getUuid(), user);
+            const message = new UpdateSpaceUserMessage();
+            message.setSpacename(this.name);
+            message.setUser(spaceUser);
+            this.notifyWatchers(watcher, message);
+            debug(`${this.name} : user => updated ${spaceUser.getUuid()}`);
+        }
+    }
+    public removeUser(watcher: SpacesWatcher, uuid: string) {
+        const usersList = this.usersList(watcher);
+        usersList.delete(uuid);
+
+        const message = new RemoveSpaceUserMessage();
+        message.setSpacename(this.name);
+        message.setUseruuid(uuid);
+        this.notifyWatchers(watcher, message);
+        debug(`${this.name} : user => removed ${uuid}`);
+    }
+
+    public addWatcher(watcher: SpacesWatcher) {
+        this.users.set(watcher, new Map<string, SpaceUser>());
+        debug(`Space ${this.name} => watcher added ${watcher.uuid}`);
+        [...this.users.values()].forEach((spaceUsers) => {
+            [...spaceUsers.values()].forEach((spaceUser) => {
+                const message = new AddSpaceUserMessage();
+                message.setSpacename(this.name);
+                message.setUser(spaceUser);
+                watcher.write(message);
+            });
+        });
+    }
+    public removeWatcher(watcher: SpacesWatcher) {
+        this.users.delete(watcher);
+        debug(`${this.name} => watcher removed ${watcher.uuid}`);
+    }
+
+    private notifyWatchers(watcher: SpacesWatcher, message: SpaceMessage) {
+        [...this.users.keys()].forEach((watcher_) => {
```

Comment:
Same thing here regarding performance. Use a `for ...  of ...`of loop without array destructuring.